File: java/code/Solution_2.java
// Solution for Problem-2 in java as follows:- 

package code;

import java.util.HashMap;
import java.util.Scanner;

public class Solution_2 {

    public int longestSubarrayWithSumZero(int[] A) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int sum = 0;
        int maxLength = 0;
        map.put(0, -1);

        for (int i = 0; i < A.length; i++) {
            sum += A[i]; 

            if (map.containsKey(sum)) {
                maxLength = Math.max(maxLength, i - map.get(sum));
            } else {
                map.put(sum, i);
            }
        }

        return maxLength;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Solution_2 solution = new Solution_2();
        int n = scanner.nextInt();

        int[] A = new int[n];

        for (int i = 0; i < n; i++) {
            A[i] = scanner.nextInt();
        } 
        System.out.println("Length of the longest subarray with sum zero: " + solution.longestSubarrayWithSumZero(A));
        scanner.close();

    }
}

File: cpp/code/Solution-2.cpp
// Solution for Problem-2 in c++ as follows:- 

#include <iostream>
#include <unordered_map>
#include <vector>
#include <algorithm>

using namespace std;

int longestSubarrayWithSumZero(vector<int>& A) {
    unordered_map<int, int> mp;
    int sum = 0, ans = 0;
    mp[0] = -1;

    for (int i = 0; i < A.size(); i++) {
        sum += A[i];
        if (mp.count(sum)) {
            ans = max(ans, i - mp[sum]);
        } else {
            mp[sum] = i;
        }
    }
    return ans;
}

int main() {
    int n;
    cin>>n;
    vector<int> A;
    for(int i=0;i<n;i++){
        cin>>A[i];
    }
    cout<<"The longest Subarray with zero Sum is: "<<longestSubarrayWithSumZero(A)<<endl;
    return 0;
}
